promql/engine.go

```diff
@@ -742,34 +724,48 @@ func extractGroupsFromPath(p []parser.Node) (bool, []string) {
 	return false, nil
 }
 
-func checkForSeriesSetExpansion(ctx context.Context, expr parser.Expr) {
+func checkAndExpandSeriesSet(ctx context.Context, expr parser.Expr) storage.Warnings {
 	switch e := expr.(type) {
 	case *parser.MatrixSelector:
-		checkForSeriesSetExpansion(ctx, e.VectorSelector)
+		if ws := checkAndExpandSeriesSet(ctx, e.VectorSelector); len(ws) > 0 {
+			return ws
+		}
 	case *parser.VectorSelector:
 		if e.Series == nil {
-			series, err := expandSeriesSet(ctx, e.UnexpandedSeriesSet)
+			series, ws, err := expandSeriesSet(ctx, e.UnexpandedSeriesSet)
 			if err != nil {
-				panic(err)
+				panic(errorWithWarnings{err: err, warnings: ws})
 			} else {
 				e.Series = series
 			}
+			if len(ws) > 0 {
+				return ws
+			}
 		}
 	}
+	return nil
 }
 
-func expandSeriesSet(ctx context.Context, it storage.SeriesSet) (res []storage.Series, err error) {
+func expandSeriesSet(ctx context.Context, it storage.SeriesSet) (res []storage.Series, ws storage.Warnings, err error) {
 	for it.Next() {
 		select {
 		case <-ctx.Done():
-			return nil, ctx.Err()
+			return nil, nil, ctx.Err()
 		default:
 		}
 		res = append(res, it.At())
+		ws = append(ws, it.Warnings()...)
 	}
-	return res, it.Err()
+	return res, ws, it.Err()
 }
 
+type errorWithWarnings struct {
```

Comment:
What is this? I don't  think we need to preserve warnings on error? cc @brian-brazil what would be your opinion? :thinking:

Reply:
We don't have to, but it wouldn't hurt.

Reply:
Cool, maybe it would be good candidate to do in following PR